runtime/inspector/mod.rs

```diff
@@ -1053,6 +717,127 @@ impl InspectorSession {
   }
 }
 
+/// A local inspector session that can be used to send and receive protocol messages directly on
+/// the same thread as an isolate.
+pub struct InMemorySession2 {
+  v8_session_tx: UnboundedSender<Result<Vec<u8>, AnyError>>,
+  v8_session_rx: UnboundedReceiver<(Option<i32>, String)>,
+  response_tx_map: HashMap<i32, oneshot::Sender<serde_json::Value>>,
+  next_message_id: i32,
+  notification_queue: Vec<Value>,
+}
+
+impl InMemorySession2 {
+  pub fn new(
+    v8_session_tx: UnboundedSender<Result<Vec<u8>, AnyError>>,
+    v8_session_rx: UnboundedReceiver<(Option<i32>, String)>,
+  ) -> Self {
+    let response_tx_map = HashMap::new();
+    let next_message_id = 0;
+
+    let notification_queue = Vec::new();
+
+    Self {
+      v8_session_tx,
+      v8_session_rx,
+      response_tx_map,
+      next_message_id,
+      notification_queue,
+    }
+  }
+
+  pub fn notifications(&mut self) -> Vec<Value> {
+    self.notification_queue.split_off(0)
+  }
+
+  pub async fn post_message(
+    &mut self,
+    method: &str,
+    params: Option<serde_json::Value>,
+  ) -> Result<serde_json::Value, AnyError> {
+    let id = self.next_message_id;
+    self.next_message_id += 1;
+
+    let (response_tx, mut response_rx) =
+      oneshot::channel::<serde_json::Value>();
+    self.response_tx_map.insert(id, response_tx);
+
+    let message = json!({
+        "id": id,
+        "method": method,
+        "params": params,
+    });
+
+    let raw_message = serde_json::to_string(&message).unwrap();
+    eprintln!("sending message from repl");
+    self
+      .v8_session_tx
+      .unbounded_send(Ok(raw_message.as_bytes().to_vec()))
+      .unwrap();
+    eprintln!("sent message from repl");
+
+    loop {
+      eprintln!("looping");
+      tokio::select! {
+        _ = self.receive_from_v8_session() => {
+          eprintln!("received vrom v8_session")
+        },
+        result = &mut response_rx => {
+          eprintln!("got response");
+          let response = result?;
+          if let Some(error) = response.get("error") {
+            return Err(generic_error(error.to_string()));
+          }
+
+          let result = response.get("result").unwrap().clone();
+          return Ok(result);
+        }
+      }
+    }
+  }
+
+  async fn receive_from_v8_session(&mut self) {
+    eprintln!("waiting on v8_session response");
+    let (maybe_call_id, message) = self.v8_session_rx.next().await.unwrap();
```

Comment:
This call never resolves and deadlocks...

Reply:
```
sending runtime enable
sending message from repl
sent message from repl
looping
waiting on v8_session response
new handshake session
Debugger session started.
got new message from proxy
sending new message to proxy {"method":"Runtime.executionContextCreated","params":{"context":{"id":1,"origin":"","name":"global context","uniqueId":"-74853213617137066.1750212901488508331"}}}
sent new message to proxy
sending new message to proxy {"id":0,"result":{}}
sent new message to proxy
```

Reply:
There's no where in the loop where it can yield?

Reply:
@caspervonb it's all async and inside a loop there's `tokio::select!` I'd expect `receive_from_v8_session` would be polled in that case. It just doesn't ever resolve; I haven't yet tracked what's the exact problem

Reply:
Should be fixed now